src/Symfony/Component/Translation/Translatable/DateTimeTranslatable.php

```diff
@@ -0,0 +1,76 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Translation\Translatable;
+
+use Symfony\Contracts\Translation\TranslatableInterface;
+use Symfony\Contracts\Translation\TranslatorInterface;
+
+/**
+ * Wrapper around PHP IntlDateFormatter for date and time
+ * The timezone from the DateTime instance is used instead of the server's timezone.
+ *
+ * Implementation of the ICU recommendation to first format advanced parameters before translation.
+ *
+ * @see https://unicode-org.github.io/icu/userguide/format_parse/messages/#format-the-parameters-separately-recommended
+ *
+ * @author Sylvain Fabre <syl.fabre@gmail.com>
+ */
+class DateTimeTranslatable implements TranslatableInterface
+{
+    private $dateTime;
+    private $dateType;
+    private $timeType;
+
+    private $formatters = [];
+
+    public function __construct(
+        \DateTimeInterface $dateTime,
+        int $dateType = \IntlDateFormatter::SHORT,
+        int $timeType = \IntlDateFormatter::SHORT
+    ) {
+        $this->dateTime = $dateTime;
+        $this->dateType = $dateType;
+        $this->timeType = $timeType;
+    }
+
+    public function trans(TranslatorInterface $translator, string $locale = null): string
+    {
+        $timezone = $this->dateTime->getTimezone();
+        $key = implode('.', [$locale, $this->dateType, $this->timeType, $timezone->getName()]);
```

Comment:
the cache key should use `$locale ?? $translator->getLocale()` to be consistent with the instantiation.